src/synapse/authProvider.js

```diff
@@ -13,16 +30,13 @@ const authProvider = {
       }),
     };
 
-    // add 'https://' to homeserver url if its missing
-    let newUrl = window.decodeURIComponent(homeserver);
-    newUrl = newUrl.trim().replace(/\s/g, "");
-    if (!/^https?:\/\//i.test(newUrl)) {
-      homeserver = `https://${newUrl}`;
-    }
+    const url = window.decodeURIComponent(homeserver);
+    const trimmed_url = url.trim().replace(/\s/g, "");
+    const homeserver_url = ensureHttpsForUrl(trimmed_url) + "/_matrix/client/r0/login";
 
-    const url = homeserver + "/_matrix/client/r0/login";
-    return fetchUtils.fetchJson(url, options).then(({ json }) => {
-      localStorage.setItem("home_server", json.home_server);
+    return fetchUtils.fetchJson(homeserver_url, options).then(({ json }) => {
+      const normalized_base_url = ensureHttpsForUrl(stripTrailingSlash(json.well_known["m.homeserver"].base_url))
```

Comment:
If the `base_url` does not start with `https://` this is probably a server misconfiguration. So I don't think we should try to be smarter than the server admin here.